sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/StorageSeekableByteChannel.java

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.common;
+
+import com.azure.core.util.logging.ClientLogger;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.NonReadableChannelException;
+import java.nio.channels.NonWritableChannelException;
+import java.nio.channels.SeekableByteChannel;
+
+/**
+ * StorageSeekableByteChannel allows for uploading and downloading data to and from an Azure Storage service using the
+ * {@link SeekableByteChannel} interface.
+ * <p>
+ * Storage channels are opened for read OR write access, not both. Not all APIs are supported depending on the channel
+ * mode and storage resource type (e.g. blob).
+ */
+public final class StorageSeekableByteChannel implements SeekableByteChannel {
+    private static final ClientLogger LOGGER = new ClientLogger(StorageSeekableByteChannel.class);
+
+    /**
+     * Interface for injectable behavior to read from a backing Storage resource.
+     */
+    public interface ReadBehavior {
+        /**
+         * Reads n bytes from the backing resource, where {@code 0 <= n <= dst.remaining()}.
+         * Emulates behavior of {@link java.nio.channels.ReadableByteChannel#read(ByteBuffer)}.
+         *
+         * @param dst Destination to read the resource into.
+         * @param sourceOffset Offset to read from the resource.
+         * @return Number of bytes read from the resource, possibly zero, or -1 end of resource.
+         * @see java.nio.channels.ReadableByteChannel#read(ByteBuffer)
+         */
+        int read(ByteBuffer dst, long sourceOffset);
```

Comment:
After reading through this PR more, I'd have these interface methods have a checked throw of `IOException` to align with the APIs in `SeekableByteChannel` to make the implementations have to do less try-catching.